### ISSUE TITLE
Suggestion:  Passing Args to Callback in Flux.train!, and return args from Loss()

### ISSUE THREAD

**Issue by cems2** [JuliaDiffEq/DiffEqFlux.jl#116]
**Problem:**  Flux.train allows for a callback but without any args.  This creates unweildy interfaces requiring global variables to work around.  A different but, as it turns out, related problem occurs with the loss functions return args.  In some problems the loss function computes expensive sub calculations one would like to access without redoing-- again demanding global variables.  

**Rationale:**
It turns out that solving these in a combined way is better than ad hoc independent solutions to both.  Rather than some rigid definition of what values get passed to the call-back we will use a free byproduct of the user-defined loss function call to create a user-defined set of args for the user-defined callback.  

**Suggested resolution:**
Allow the loss function to return a tuple with two args.  The first arg is the standard scalar loss required by Flux to optimize.  The second arg is any user-defined payload-- for example, a tuple of intermediate calculations in the loss functions such as the last output of the network and separate loss terms like regularizers.  

This second arg is to be ignored by Flux.train! itself, but will be passed as the arg to the designated callback function.

Example:
```
function loss(x)   **# this is a typical looking loss function, nothing special at all**
   pred = prediction(x.value)
   loss1  = abs2.( pred - x.labels)
   r = regularizers(x.network)
   totalLoss = sum( loss1) + r
   payload = ( loss1, pred, r)
  return totalLoss, payload
end

function callback(totalLoss, payload)  # the callback takes an arg
   println ("totalLoss", totalLoss ) 
   println ("Current Regulaizer value ",payload[3]
   println ("Current total Loss ", payload[1] )
   Plot( payload[2] )  **# plot the details of the prediction without recomputing it!** 
   # maybe some decision criteria to save a copy of the best loss or stopping criteria
end

# the loss train! function retains the same interface.
Flux.train!( loss, parameters, dataset, opt; cb = callback )

```

**maintaining backward compatibility:**
All the new train! function has to do is detect if loss returned a single scalar or a tuple.  If a tuple it extracts the loss for it's own use and then passes the tuple  to callback. if a scaler it calls the callback with no args.

**Comment by cems2** [comment]
I am moving this to Flux.jl issues
https://github.com/FluxML/Flux.jl/issues/999
